plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/FileSystemExchangeSource.java

```diff
@@ -77,54 +91,44 @@ public synchronized Slice read()
             return null;
         }
 
-        if (sliceInput != null && !sliceInput.isReadable()) {
-            sliceInput.close();
-            sliceInput = null;
-        }
+        Iterator<ExchangeStorageReader> iterator = readers.iterator();
 
-        if (sliceInput == null) {
-            if (files.hasNext()) {
-                // TODO: implement parallel read
-                URI file = files.next();
-                Optional<SecretKey> secretKey = secretKeys.next();
+        while (iterator.hasNext()) {
+            ExchangeStorageReader reader = iterator.next();
+
+            if (reader.isBlocked().isDone()) {
+                Slice slice;
                 try {
-                    sliceInput = exchangeStorage.getSliceInput(file, secretKey);
+                    slice = reader.read();
                 }
                 catch (IOException e) {
                     throw new UncheckedIOException(e);
                 }
+                if (slice == null) {
+                    verify(!exchangeSourceFileIterator.hasNext(), "null shouldn't be returned when there are still files to read");
+                    iterator.remove();
+                }
+                else {
+                    return slice;
+                }
             }
         }
 
-        if (sliceInput == null) {
-            return null;
-        }
-
-        if (!sliceInput.isReadable()) {
-            sliceInput.close();
-            sliceInput = null;
-            return null;
-        }
-
-        int size = sliceInput.readInt();
-        return sliceInput.readSlice(size);
+        return null;
     }
 
     @Override
     public synchronized long getMemoryUsage()
     {
-        return sliceInput != null ? sliceInput.getRetainedSize() : 0;
+        return readers.stream().mapToLong(ExchangeStorageReader::getRetainedSize).sum();
```

Comment:
Since `readers` is immutable it doesn't have to be synchronized (assuming `ExchangeStorageReader` is thread safe). Also maybe use a simple loop for efficiency